File: frontend/src/types/analysis.ts
/**
 * Analysis result type definitions.
 */

import type { ExclusionInfo } from './employee'

export type AnalysisStatus = 'PASS' | 'RISK' | 'FAIL' | 'ERROR'
export type ViewMode = 'PASS_FAIL' | 'MARGIN' | 'RISK_ZONE'

/**
 * Rate value as a decimal fraction (0.0 to 1.0).
 * Backend API expects and returns rates in this format.
 * Example: 0.06 represents 6%, 0.75 represents 75%.
 */
export type DecimalRate = number

export interface ScenarioResult {
  status: AnalysisStatus
  nhce_acp: number | null
  hce_acp: number | null
  limit_125?: number | null
  limit_2pct_uncapped?: number | null
  cap_2x?: number | null
  limit_2pct_capped?: number | null
  effective_limit?: number | null
  max_allowed_acp: number | null
  margin: number | null
  binding_rule?: '1.25x' | '2pct/2x' | null
  /** Adoption rate as decimal (0.0-1.0) */
  adoption_rate: DecimalRate
  /** Contribution rate as decimal (0.0-1.0) */
  contribution_rate: DecimalRate
  seed_used: number
  excluded_count?: number | null
  exclusion_breakdown?: ExclusionInfo | null
}

export interface GridSummary {
  pass_count: number
  risk_count: number
  fail_count: number
  error_count: number
  total_count: number
  first_failure_point?: {
    /** Adoption rate as decimal (0.0-1.0) */
    adoption_rate: DecimalRate
    /** Contribution rate as decimal (0.0-1.0) */
    contribution_rate: DecimalRate
  }
  /** Max safe contribution rate as decimal (0.0-1.0) */
  max_safe_contribution?: DecimalRate
  worst_margin?: number
  excluded_count?: number
  exclusion_breakdown?: ExclusionInfo | null
}

export interface GridResult {
  scenarios: ScenarioResult[]
  summary: GridSummary
  seed_used: number
}

// RunDetail is defined in runService.ts to avoid circular dependency

File: backend/app/models/analysis.py
"""Analysis result Pydantic models."""

from __future__ import annotations

from enum import Enum
from typing import Optional

from pydantic import BaseModel


class AnalysisStatus(str, Enum):
    """Scenario analysis status."""

    PASS = "PASS"
    RISK = "RISK"
    FAIL = "FAIL"
    ERROR = "ERROR"


class ScenarioResult(BaseModel):
    """Result of a single scenario analysis."""

    status: AnalysisStatus
    nhce_acp: Optional[float] = None
    hce_acp: Optional[float] = None
    limit_125: Optional[float] = None
    limit_2pct_uncapped: Optional[float] = None
    cap_2x: Optional[float] = None
    limit_2pct_capped: Optional[float] = None
    effective_limit: Optional[float] = None
    max_allowed_acp: Optional[float] = None
    margin: Optional[float] = None
    binding_rule: Optional[str] = None
    adoption_rate: float
    contribution_rate: float
    seed_used: int


class GridSummary(BaseModel):
    """Aggregate metrics for grid analysis."""

    pass_count: int
    risk_count: int
    fail_count: int
    error_count: int
    total_count: int
    first_failure_point: Optional[dict] = None
    max_safe_contribution: Optional[float] = None
    worst_margin: Optional[float] = None


class GridResult(BaseModel):
    """Collection of scenario results."""

    scenarios: list[ScenarioResult]
    summary: GridSummary
    seed_used: int


class RunDetail(BaseModel):
    """Run with full results."""

    # Include all Run fields plus results
    id: str
    workspace_id: str
    name: Optional[str] = None
    adoption_rates: list[float]
    contribution_rates: list[float]
    seed: int
    status: str
    created_at: str
    completed_at: Optional[str] = None
    results: Optional[GridResult] = None
